content/blog/2022-05-03-e2e-cv-pt1.md

```diff
@@ -0,0 +1,307 @@
+picture: 2022-05-03/e2e-cv-pt1-cover.png
+author: alex_kim
+commentsUrl: https://discuss.dvc.org/t/end-to-end-computer-vision/1178
+tags:
+  - Computer Vision
+  - DVC
+  - CML
+  - Studio
+  - CI/CD
+  - Experiment Tracking
+---
+
+### Introduction
+
+A typical ML project lifecycle goes through three phases:
+
+1. Active exploration or proof-of-concept phase. Here we try many different
+   approaches to data preprocessing, model architectures, and hyperparameter
+   tuning with the goal of finally settling on the most promising combination.
+2. Application development phase. Here we build all the “plumbing” around the
+   model, that is: getting input data, massaging it into the right format,
+   passing it to the model, and, finally, serving the model's output. The goal
+   here is to end up with a version of an end-to-end application that is robust
+   and well-performing enough to be considered for deployment to the production
+   environment.
+3. Production deployment phase. Here we promote a well-tested version of the
+   application from the development environment to the production environment.
+
+In this series of posts we'll describe an approach that streamlines all three
+phases. As our demo project, we've selected a very common deployment pattern in
+CV projects: a CV model wrapped in a web API service. Automatic defect detection
+is an example problem that we are addressing with this pattern. Specifically,
+we'll perform an
+[image segmentation](https://ai.stanford.edu/~syyeung/cvweb/tutorial3.html) task
+on a magnetic tiles dataset first introduced in this
+[paper](https://www.researchgate.net/profile/Congying-Qiu/publication/327701995_Saliency_defect_detection_of_magnetic_tiles/links/5b9fd1bd45851574f7d25019/Saliency-defect-detection-of-magnetic-tiles.pdf)
+and available in this GitHub
+[repository](https://github.com/abin24/Magnetic-tile-defect-datasets.).
+
+![A sample from the image segmentation dataset we used for this project. Top row: images of magnetic tile surfaces. Bottom row: segmentation mask (white pixels show defective areas)](/uploads/images/2022-05-03/dataset_sample.png '=800')
+
+- This post (part 1) introduces the concepts of data versioning and ML pipelines
+  as they apply to Computer Vision projects.
+- Part 2 will focus on experiment tracking and management - key components
+  needed for effective collaboration between team members.
+- In part 3, you’ll learn how to easily move your model training workloads from
+  a local machine to cloud infrastructure and set up proper CI/CD workflows for
+  ML projects.
+
+### Target Audience
+
+We assume the target audience of this post to be technical folks (e.g. Software
+Engineers, ML Engineers, Data Scientists) who are familiar with the general
+Machine Learning and Computer Vision concepts, Python programming, CI/CD
```

Comment:
I'm a Software Engineer, not familiar with CV though. And I'm having a hard time understanding the abstract and intro.

Reply:
Slightly rephrased.